From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Sat, 20 Jul 2019 16:59:51 +0300
Subject: [PATCH] Various micro-optimizations

* Optimize query packet string writing
* Improve ProxyServer#matchPlayer
* Don't use Class#getSimpleName()
* Don't initialize reconnect save thread when it's not set
* Add ServerPing serializer
* Re-use static chat components instead of reallocating them

diff --git a/api/src/main/java/eu/mikroskeem/firefly/api/misc/ClassUtil.java b/api/src/main/java/eu/mikroskeem/firefly/api/misc/ClassUtil.java
new file mode 100644
index 0000000000000000000000000000000000000000..546bef8e2a62c307c2e59dce884489148b23722f
--- /dev/null
+++ b/api/src/main/java/eu/mikroskeem/firefly/api/misc/ClassUtil.java
@@ -0,0 +1,21 @@
+package eu.mikroskeem.firefly.api.misc;
+
+/**
+ * @author Mark Vainomaa
+ */
+public final class ClassUtil {
+    private ClassUtil() {}
+
+    public static String getClassName(Class<?> clazz) {
+        return getClassNameFromString(clazz.getName());
+    }
+
+    public static String getClassName(Object instance) {
+        return getClassNameFromString(instance.getClass().getName());
+    }
+
+    public static String getClassNameFromString(String className) {
+        int i = className.lastIndexOf('.');
+        return i > 0 ? className.substring(i + 1) : className;
+    }
+}
diff --git a/api/src/main/java/net/md_5/bungee/Util.java b/api/src/main/java/net/md_5/bungee/Util.java
index 888cfe2f52236328a95a17aec11df5b5d24aaa5b..3d0e4b3ebba0757ee336b9f39f82145eda5c2387 100644
--- a/api/src/main/java/net/md_5/bungee/Util.java
+++ b/api/src/main/java/net/md_5/bungee/Util.java
@@ -82,7 +82,7 @@ public class Util
     {
         // TODO: We should use clear manually written exceptions
         StackTraceElement[] trace = t.getStackTrace();
-        return t.getClass().getSimpleName() + " : " + t.getMessage()
+        return eu.mikroskeem.firefly.api.misc.ClassUtil.getClassName(t.getClass()) + " : " + t.getMessage() // Firefly - don't use Class#getSimpleName()
                 + ( ( trace.length > 0 ) ? " @ " + t.getStackTrace()[0].getClassName() + ":" + t.getStackTrace()[0].getLineNumber() : "" );
     }
 
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftDecoder.java b/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftDecoder.java
index 9333b64248c4ab89fe9ce2829dbf09039bdfc064..9615de6e74f95abf962177c3ce6e114bb7864687 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftDecoder.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftDecoder.java
@@ -29,6 +29,7 @@ public class MinecraftDecoder extends MessageToMessageDecoder<ByteBuf>
     @Setter private static Runnable emptyPacketMetric;
     @Setter private static Runnable invalidPacketMetric;
     // Firefly end
+    @Setter private static java.util.function.Function<Class<?>, String> getClassNameFunction; // Firefly - don't use Class#getSimpleName()
 
     public MinecraftDecoder(Protocol protocol, boolean server, int protocolVersion) {
         this.protocol = protocol;
@@ -80,7 +81,7 @@ public class MinecraftDecoder extends MessageToMessageDecoder<ByteBuf>
             if (packetTypeInfo instanceof Integer) {
                 packetTypeStr = "id " + Integer.toHexString((Integer) packetTypeInfo);
             } else if (packetTypeInfo instanceof Class) {
-                packetTypeStr = "class " + ((Class) packetTypeInfo).getSimpleName();
+                packetTypeStr = "class " + getClassNameFunction.apply((Class) packetTypeInfo); // Firefly - don't use Class#getSimpleName()
             } else {
                 packetTypeStr = "unknown";
             }
diff --git a/proxy/src/main/java/eu/mikroskeem/firefly/gson/ServerPingSerializer.java b/proxy/src/main/java/eu/mikroskeem/firefly/gson/ServerPingSerializer.java
new file mode 100644
index 0000000000000000000000000000000000000000..29e05a380ce3e0ee638c8297437dc1b9c5879ae9
--- /dev/null
+++ b/proxy/src/main/java/eu/mikroskeem/firefly/gson/ServerPingSerializer.java
@@ -0,0 +1,35 @@
+package eu.mikroskeem.firefly.gson;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonSerializationContext;
+import com.google.gson.JsonSerializer;
+import net.md_5.bungee.api.ServerPing;
+
+import java.lang.reflect.Type;
+
+/**
+ * @author Mark Vainomaa
+ */
+public final class ServerPingSerializer implements JsonSerializer<ServerPing> {
+    @Override
+    public JsonElement serialize(ServerPing src, Type typeOfSrc, JsonSerializationContext context) {
+        JsonObject root = new JsonObject();
+        root.add("description", context.serialize(src.getDescriptionComponent()));
+        root.add("favicon", context.serialize(src.getFaviconObject()));
+        root.add("modinfo", context.serialize(src.getModinfo()));
+
+        JsonObject version = new JsonObject();
+        version.addProperty("name", src.getVersion().getName());
+        version.addProperty("protocol", src.getVersion().getProtocol());
+        root.add("version", version);
+
+        JsonObject players = new JsonObject();
+        players.addProperty("max", src.getPlayers().getMax());
+        players.addProperty("online", src.getPlayers().getOnline());
+        players.add("sample", context.serialize(src.getPlayers().getSample()));
+        root.add("players", players);
+
+        return root;
+    }
+}
diff --git a/proxy/src/main/java/io/github/waterfallmc/waterfall/console/WaterfallConsole.java b/proxy/src/main/java/io/github/waterfallmc/waterfall/console/WaterfallConsole.java
index 6cec0b5a141a41e9f4e36c25f6f7460aa5dacdc2..0dfde6a2e331f4d4166ec6d7ee0b87790968e82f 100644
--- a/proxy/src/main/java/io/github/waterfallmc/waterfall/console/WaterfallConsole.java
+++ b/proxy/src/main/java/io/github/waterfallmc/waterfall/console/WaterfallConsole.java
@@ -9,6 +9,7 @@ import org.jline.reader.LineReader;
 import org.jline.reader.LineReaderBuilder;
 
 public final class WaterfallConsole extends SimpleTerminalConsole {
+    private static final net.md_5.bungee.api.chat.BaseComponent[] COMMAND_NOT_FOUND = new ComponentBuilder("Command not found").color(ChatColor.RED).create(); // Firefly - micro-optimize
 
     @Override
     protected LineReader buildReader(LineReaderBuilder builder) {
@@ -28,7 +29,7 @@ public final class WaterfallConsole extends SimpleTerminalConsole {
     protected void runCommand(String command) {
         ProxyServer proxy = ProxyServer.getInstance();
         if (!proxy.getPluginManager().dispatchCommand(proxy.getConsole(), command)) {
-            proxy.getConsole().sendMessage(new ComponentBuilder("Command not found").color(ChatColor.RED).create());
+            proxy.getConsole().sendMessage(COMMAND_NOT_FOUND);//new ComponentBuilder("Command not found").color(ChatColor.RED).create()); // Firefly - micro-optimize
         }
     }
 
diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index 7f13eea6bf7245f2b61af584c95c537604be23ca..cbcc3bad233b95bab1728dcf82911cdc2058f05c 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -119,7 +119,7 @@ public class BungeeCord extends ProxyServer
     /**
      * locations.yml save thread.
      */
-    private final Timer saveThread = new Timer( "Reconnect Saver" );
+    private /*final*/ Timer saveThread /*= new Timer( "Reconnect Saver" )*/; // Firefly - don't initialize reconnect save thread when it's not set
     // private final Timer metricsThread = new Timer( "Metrics Thread" ); // Waterfall: Disable Metrics
     /**
      * Server socket listener.
@@ -169,6 +169,7 @@ public class BungeeCord extends ProxyServer
             .registerTypeAdapter( KeybindComponent.class, new KeybindComponentSerializer() )
             .registerTypeAdapter( ScoreComponent.class, new ScoreComponentSerializer() )
             .registerTypeAdapter( SelectorComponent.class, new SelectorComponentSerializer() )
+            .registerTypeAdapter(ServerPing.class, new eu.mikroskeem.firefly.gson.ServerPingSerializer()) // Firefly - add ServerPing serializer
             .registerTypeAdapter( ServerPing.PlayerInfo.class, new PlayerInfoSerializer() )
             .registerTypeAdapter( Favicon.class, Favicon.getFaviconTypeAdapter() ).create();
     @Getter
@@ -353,9 +354,13 @@ public class BungeeCord extends ProxyServer
         {
             connectionThrottle = new ConnectionThrottle( config.getThrottle(), config.getThrottleLimit() );
         }
+        net.md_5.bungee.protocol.MinecraftDecoder.setGetClassNameFunction(eu.mikroskeem.firefly.api.misc.ClassUtil::getClassName); // Firefly - don't use Class#getSimpleName()
         eu.mikroskeem.firefly.misc.PrometheusMetrics.initialize(); // Firefly - add Prometheus metrics support
         startListeners();
 
+        // Firefly start - don't initialize reconnect save thread when it's not set
+        if (getReconnectHandler() != null && (saveThread = new Timer( "Reconnect Saver" )) != null) // Last "check" is intended and does basically a lazy init
+        // Firefly end
         saveThread.scheduleAtFixedRate( new TimerTask()
         {
             @Override
@@ -547,6 +552,7 @@ public class BungeeCord extends ProxyServer
             reconnectHandler.save();
             reconnectHandler.close();
         }
+                if (saveThread != null) // Firefly - don't initialize reconnect save thread when it's not set
         saveThread.cancel();
         //metricsThread.cancel(); // Waterfall: Disable Metrics
 
@@ -905,6 +911,7 @@ public class BungeeCord extends ProxyServer
             return Collections.singleton( exactMatch );
         }
 
+        /* Firefly start - improve matchPlayer
         return Sets.newHashSet( Iterables.filter( getPlayers(), new Predicate<ProxiedPlayer>()
         {
 
@@ -914,6 +921,19 @@ public class BungeeCord extends ProxyServer
                 return ( input == null ) ? false : input.getName().toLowerCase( Locale.ROOT ).startsWith( partialName.toLowerCase( Locale.ROOT ) );
             }
         } ) );
+        */
+        String lowerPartialMatch = partialName.toLowerCase(Locale.ROOT);
+        var foundPlayers = new java.util.LinkedList<ProxiedPlayer>();
+        for (var player : getPlayers()) {
+            if (player == null) { // Should not happen though?
+                continue;
+            }
+            if (player.getName().toLowerCase(Locale.ROOT).startsWith(lowerPartialMatch)) {
+                foundPlayers.add(player);
+            }
+        }
+        return java.util.Set.copyOf(foundPlayers);
+        // Firefly end
     }
 
     @Override
diff --git a/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java b/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java
index 989bfd87b9e6d62abba28d51984b8526d6298bfd..2cf0d588eb78fb724bf7a15f4656dfd1d92ad10d 100644
--- a/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java
+++ b/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java
@@ -156,7 +156,7 @@ public class HandlerBoss extends ChannelInboundHandlerAdapter
                 {
                     ProxyServer.getInstance().getLogger().log( Level.WARNING, "{0} - {1}: {2}", new Object[]
                     {
-                        handler, cause.getClass().getSimpleName(), cause.getMessage()
+                        handler, eu.mikroskeem.firefly.api.misc.ClassUtil.getClassName(cause), cause.getMessage() // Firefly - don't use Class#getSimpleName()
                     } );
                 } else if ( cause instanceof QuietException )
                 {
diff --git a/query/src/main/java/net/md_5/bungee/query/QueryHandler.java b/query/src/main/java/net/md_5/bungee/query/QueryHandler.java
index 1fe520b31a58483b6aae920458d3cb968b25e6b1..40b76f48b515f6660d0eaa46f9e01289e7d85c31 100644
--- a/query/src/main/java/net/md_5/bungee/query/QueryHandler.java
+++ b/query/src/main/java/net/md_5/bungee/query/QueryHandler.java
@@ -48,10 +48,14 @@ public class QueryHandler extends SimpleChannelInboundHandler<DatagramPacket>
 
     private void writeString(ByteBuf buf, String s)
     {
+        /* // Firefly start - optimize query packet string writing
         for ( char c : s.toCharArray() )
         {
             buf.writeByte( c );
         }
+        */
+        buf.writeCharSequence(s, java.nio.charset.StandardCharsets.ISO_8859_1);
+        // Firefly end
         buf.writeByte( 0x00 );
     }
 
